File: build-logic/java-plugins/src/main/java/com/nisecoder/helloworld/gradle/KotlinPlugin.java
package com.nisecoder.helloworld.gradle;

import org.gradle.api.JavaVersion;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.tasks.testing.Test;
import org.gradle.jvm.toolchain.JavaLanguageVersion;
import org.gradle.jvm.toolchain.JavaToolchainService;
import org.gradle.jvm.toolchain.JvmVendorSpec;
import org.jetbrains.kotlin.gradle.dsl.KotlinJvmProjectExtension;
import org.jetbrains.kotlin.gradle.tasks.KotlinJvmCompile;

public class KotlinPlugin implements Plugin<Project> {
    @Override
    public void apply(Project project) {
        project.getPlugins().apply("org.jetbrains.kotlin.jvm");

        JavaToolchainService javaToolchain = (JavaToolchainService) project.getExtensions().getByName("javaToolchains");
        javaToolchain.compilerFor(spec -> {
            spec.getLanguageVersion().set(JavaLanguageVersion.of(17));
            spec.getVendor().set(JvmVendorSpec.ADOPTIUM);
        });

        project.getExtensions().configure(KotlinJvmProjectExtension.class, it ->
            it.jvmToolchain(spec -> spec.getLanguageVersion().set(JavaLanguageVersion.of(17)))
        );

        var tasks = project.getTasks();
        tasks.withType(KotlinJvmCompile.class).configureEach(task -> {
            var options = task.getKotlinOptions();
            options.setLanguageVersion("1.8");
            options.setApiVersion("1.8");
            options.setJvmTarget(JavaVersion.VERSION_17.toString());
            options.setJavaParameters(true);
        });

        var dependencies = project.getDependencies();
        dependencies.add("implementation", dependencies.platform("org.jetbrains.kotlin:kotlin-bom:1.7.10"));
        dependencies.add("testImplementation", dependencies.platform("org.jetbrains.kotlin:kotlin-bom:1.7.10"));
        dependencies.add("testImplementation", "org.jetbrains.kotlin:kotlin-test");

        Test test = (Test) tasks.getByName("test");
        var processorNumber = Runtime.getRuntime().availableProcessors();
        if (processorNumber > 0) {
            test.setMaxParallelForks(processorNumber);
        } else {
            test.setMaxParallelForks(1);
        }

        test.reports(reports -> {
            var rootBuildDir = project.getRootProject().getLayout().getBuildDirectory();
            var junitXml = reports.getJunitXml();
            junitXml.getRequired().set(true);
            junitXml.getOutputLocation().set(rootBuildDir.dir("test-results/junit/" + project.getName()));

            var html = reports.getHtml();
            html.getRequired().set(true);
            html.getOutputLocation().set(rootBuildDir.dir("reports/tests/test/" + project.getName()));
        });
    }
}

File: build-logic/java-plugins/src/main/kotlin/com/nisecoder/helloworld/gradle/KotlinPlugin.kt
package com.nisecoder.helloworld.gradle

import org.gradle.api.JavaVersion
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.tasks.testing.Test
import org.gradle.jvm.toolchain.JavaLanguageVersion
import org.gradle.jvm.toolchain.JavaToolchainService
import org.gradle.jvm.toolchain.JavaToolchainSpec
import org.gradle.jvm.toolchain.JvmVendorSpec
import org.gradle.kotlin.dsl.apply
import org.gradle.kotlin.dsl.configure
import org.gradle.kotlin.dsl.getByName
import org.gradle.kotlin.dsl.kotlin
import org.gradle.kotlin.dsl.withType
import org.jetbrains.kotlin.gradle.dsl.KotlinJvmProjectExtension
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile

class KotlinPlugin: Plugin<Project> {
    override fun apply(project: Project) : Unit = project.run {
        plugins.apply("org.jetbrains.kotlin.jvm")
        plugins.apply(CodeQualityPlugin::class)

        extensions.getByName<JavaToolchainService>("javaToolchains").compilerFor {
                @Suppress("MagicNumber")
                languageVersion.set(JavaLanguageVersion.of(11))
            @Suppress("UnstableApiUsage")
            vendor.set(JvmVendorSpec.ADOPTIUM)
        }

        configure<KotlinJvmProjectExtension> {
            jvmToolchain {
                languageVersion.set(JavaLanguageVersion.of(11))
            }
        }

        tasks.withType<KotlinCompile>().configureEach {
            kotlinOptions {
                languageVersion = "1.7"
                apiVersion = "1.7"
                jvmTarget = JavaVersion.VERSION_11.toString()
                javaParameters = true
            }
        }

        dependencies.apply {
            add("implementation", platform("org.jetbrains.kotlin:kotlin-bom:1.7.10"))
            add("testImplementation", platform("org.jetbrains.kotlin:kotlin-bom:1.7.10"))
            add("testImplementation", kotlin("test"))
        }

        tasks.getByName<Test>("test") {
            maxParallelForks = (Runtime.getRuntime().availableProcessors() / 2).takeIf { it > 0 } ?: 1

            reports {
                junitXml.apply {
                    required.set(true)
                    outputLocation.set(rootProject.layout.buildDirectory.dir("test-results/junit/${project.name}"))
                }
                html.apply {
                    required.set(true)
                    outputLocation.set(rootProject.layout.buildDirectory.dir("reports/tests/test/${project.name}"))
                }
            }
        }
    }
}
